### PR TITLE
Adds transformation to convert a batch of ranges into simd types.

### PR DESCRIPTION
**Pull request by rrahn**

### CODE REVIEW

File: include/seqan3/core/simd/simd_algorithm.hpp [hunk 1]
```diff
@@ -36,6 +39,60 @@ constexpr simd_t iota_impl(scalar_t const offset, std::integer_sequence<scalar_t
     return simd_t{static_cast<scalar_t>(offset + I)...};
 }
 
+/*!\brief Transforms a batch of ranges into simd types (SOA) and writes them into the output iterator.
```
Comment (h-2) [inline include/seqan3/core/simd/simd_algorithm.hpp]: What's SOA?

File: include/seqan3/core/simd/simd_algorithm.hpp [hunk 2]
```diff
@@ -36,6 +39,60 @@ constexpr simd_t iota_impl(scalar_t const offset, std::integer_sequence<scalar_t
...
+//!\endcond
+constexpr void transform_batch_to_soa(output_iter_t out, range_t && seq_rng)
+{
+    // Check if the length is identical in debug mode.
+    if constexpr (std::ranges::SizedRange<range_t>)
+        assert(std::ranges::size(seq_rng) == static_cast<size_t>(simd_traits<simd_t>::length));
+
+    // stack memory for the cached iterator - sentinel pairs.
+    using it_pair_t = std::pair<std::ranges::iterator_t<value_type_t<range_t>>,
+                                std::ranges::iterator_t<value_type_t<range_t>>>;
```
Comment (h-2) [inline include/seqan3/core/simd/simd_algorithm.hpp]: sentinel_t

also reference_t<range_t>, right?

File: include/seqan3/core/simd/simd_algorithm.hpp [hunk 3]
```diff
@@ -36,6 +39,60 @@ constexpr simd_t iota_impl(scalar_t const offset, std::integer_sequence<scalar_t
...
+    for (auto && [rng, idx] : std::view::zip(seq_rng, std::view::iota(0)))
+        iter_cache[idx] = {std::ranges::begin(rng), std::ranges::end(rng)};
+
+    // Iterate over the length of the array or the maximal size of the underlying range.
+    // We assume all ranges have the same length.
+    simd_t simd{};
+    for (unsigned j = 0; (j < iter_cache.size()) || (iter_cache[j].first == iter_cache[j].second); ++j, ++out)
+    {
+        // Fill the simd value with the ranks of the respective alphabets.
+        for (unsigned i = 0; i < iter_cache.size(); ++i)
```
Comment (h-2) [inline include/seqan3/core/simd/simd_algorithm.hpp]: size_t instead of unsigned

File: include/seqan3/core/simd/simd_algorithm.hpp [hunk 4]
```diff
@@ -36,6 +39,60 @@ constexpr simd_t iota_impl(scalar_t const offset, std::integer_sequence<scalar_t
     return simd_t{static_cast<scalar_t>(offset + I)...};
 }
 
+/*!\brief Transforms a batch of ranges into simd types (Structure-of-Arrays) and writes them into the output iterator.
+ * \tparam simd_t        The simd type; must model seqan3::simd::simd_concept.
+ * \tparam output_iter_t The output iterator type; must model std::OutputIterator with `simd_t`.
```
Comment (smehringer) [inline include/seqan3/core/simd/simd_algorithm.hpp]: why not std::ranges::OutputIterator?

File: include/seqan3/core/simd/simd_algorithm.hpp (hunk 1 above)
Comment (rrahn) [reply include/seqan3/core/simd/simd_algorithm.hpp]: Structure of Arrays

File: include/seqan3/core/simd/simd_algorithm.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/core/simd/simd_algorithm.hpp]: Because the iterator concepts are not defined in the std::ranges namespace, are they?

Comment (smehringer) [reply include/seqan3/core/simd/simd_algorithm.hpp]: you are right

File: include/seqan3/core/simd/simd_algorithm.hpp [hunk 5]
```diff
@@ -36,6 +39,60 @@ constexpr simd_t iota_impl(scalar_t const offset, std::integer_sequence<scalar_t
...
+        assert(std::ranges::size(seq_rng) == static_cast<size_t>(simd_traits<simd_t>::length));
+
+    // stack memory for the cached iterator - sentinel pairs.
+    using it_pair_t = std::pair<std::ranges::iterator_t<reference_t<range_t>>,
+                                std::ranges::sentinel_t<reference_t<range_t>>>;
+    std::array<it_pair_t, simd_traits<simd_t>::length> iter_cache;
+
+    // Cache the iterator, sentinel pairs of the underlying ranges.
+    for (auto && [rng, idx] : std::view::zip(seq_rng, std::view::iota(0)))
+        iter_cache[idx] = {std::ranges::begin(rng), std::ranges::end(rng)};
```
Comment (marehr) [inline include/seqan3/core/simd/simd_algorithm.hpp]: We should discuss this. This should be the input and you should not do this extra step in here.

Comment (rrahn) [reply include/seqan3/core/simd/simd_algorithm.hpp]: Ok, let's discuss this! Maybe today?

File: test/unit/core/simd/simd_algorithm_test.cpp [hunk 6]
```diff
@@ -38,3 +41,52 @@ TEST(simd_algorithm, iota)
...
+
+    SIMD_EQ(out_vec[4], simd_t{0, 1, 2, 3});
+    SIMD_EQ(out_vec[5], simd_t{0, 1, 2, 3});
+    SIMD_EQ(out_vec[6], simd_t{0, 1, 2, 3});
+    SIMD_EQ(out_vec[7], simd_t{0, 1, 2, 3});
+
+    data[0] = std::span{std::ranges::data(seq1) + 2 * simd_traits<simd_t>::length, 2};
+    data[1] = std::span{std::ranges::data(seq2) + 2 * simd_traits<simd_t>::length, 2};
+    data[2] = std::span{std::ranges::data(seq3) + 2 * simd_traits<simd_t>::length, 2};
+    data[3] = std::span{std::ranges::data(seq4) + 2 * simd_traits<simd_t>::length, 2};
```
[the remark this answers is not in the document]
Comment (rrahn) [reply test/unit/core/simd/simd_algorithm_test.cpp]: nice catch

**Comment by marehr** [comment]
@rrahn I think you pushed in the wrong branch

**Comment by rrahn** [comment]
@marehr no I wasn't aware, that I added already commit from this branch. Anyway, we need to discuss this. Then I will resolve it inot new PRs

**Comment by rrahn** [comment]
I close this for now, and will hack the vectorised alignment kernel. Then I open PRs with separate components.

**Pull request closed without merging**